docs/site/migration/models/mixins.md

```diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+      super(...args);
+    }
+  }
+
+  return MixedModel;
+}
+```
+
+{% include note.html content="At the moment, [TypeScript does not allow decorators in class expressions](https://github.com/microsoft/TypeScript/issues/7342). This is why we need to declare the class with a name, and then return it." %}
+
+### Generating A Model Via The CLI
+
+A CLI-generated model named `Note` with 3 properties: **id**, **title**, and
+**content** would look like this:
+
+{% include code-caption.html content="src/models/note.model.ts" %}
+
+```ts
+import {Entity, model, property} from '@loopback/repository';
+
+@model()
+export class Note extends Entity {
+  @property({
+    type: 'number',
+    id: true,
+    generated: true,
+  })
+  id?: number;
+
+  @property({
+    type: 'string',
+    required: true,
+  })
+  title: string;
+
+  @property({
+    type: 'string',
+  })
+  content?: string;
+
+  constructor(data?: Partial<Note>) {
+    super(data);
+  }
+}
+
+export interface NoteRelations {
+  // describe navigational properties here
+}
+
+export type NoteWithRelations = Note & NoteRelations;
+```
+
+### Adjusting The Model File To Use addCategoryPropertyMixin
+
+Before we use the mixin class factory function:
+
+- rename this class to `TempNote`
+- remove the class' `export` directive
+- remove its **@model()** decorator
+- Leave the `NoteRelations` interface and the `NoteWithRelations` type alone for
+  now.
+
+{% include code-caption.html content="src/models/note.model.ts" %}
+
+```ts
+class TempNote extends Entity {
+  @property({
+    type: 'number',
+    id: true,
+    generated: true,
+  })
+  id?: number;
+
+  @property({
+    type: 'string',
+    required: true,
+  })
+  title: string;
+
+  @property({
+    type: 'string',
+  })
+  content?: string;
+
+  constructor(data?: Partial<TempNote>) {
+    super(data);
+  }
+}
+
+export interface NoteRelations {
+  // describe navigational properties here
+}
+
+export type NoteWithRelations = Note & NoteRelations;
+```
+
+Then,
+
+- import the `addCategoryPropertyMixin` mixin
+- declare a class `Note` which extends the class returned from the mixin
+  function that takes in the `TempNote` superclass an input
+- add the `export` directive and the **@model()** decorator to the `Note` class
+- move the `NoteRelations` interface and the `NoteWithRelations` type to the
+  bottom
+
+{% include code-caption.html content="src/models/note.model.ts" %}
+
+```ts
+import {Entity, model, property} from '@loopback/repository';
+import {addCategoryPropertyMixin} from '../mixins/categoryPropertyMixin';
+
+class TempNote extends Entity {
+  @property({
+    type: 'number',
+    id: true,
+    generated: true,
+  })
+  id?: number;
+
+  @property({
+    type: 'string',
+    required: true,
+  })
+  title: string;
+
+  @property({
+    type: 'string',
+  })
+  content?: string;
+
+  constructor(data?: Partial<TempNote>) {
+    super(data);
+  }
+}
+
+@model()
+export class Note extends addCategoryPropertyMixin(TempNote) {
```

Comment:
```suggestion
export class Note extends addCategoryPropertyMixin(Entity) {
```
Can you shortly explain why there is a need to rename and use `TempNote` instead of using a fully implemented `Note` extending `Entity` which was exhanced by the mixin?

Reply:
+1, I wanted to suggest the same. Take a look at how we are using mixins for Application classes:

https://github.com/strongloop/loopback-next/blob/28ebb8b7276577494d398423104d124b80241d43/examples/todo/src/application.ts#L15-L17

Ideally, I'd like use to describe the same approach for model mixins too. Here is an illustration of what I mean:

```ts
@model()
export class Note extends addCategoryPropertyMixin(Entity) {
  @property({
    type: 'number',
    id: true,
    generated: true,
  })
  id?: number;

  // etc.
}
```

Unless there are reasons why that does not work and the mixin must be applied on the class containing other model properties?

Reply:
The mixin requires something with a constructor. And `Entity` is an abstract class.

```
export abstract class Entity extends Model implements Persistable 
```
https://github.com/strongloop/loopback-next/blob/master/packages/repository/src/model.ts#L276

Reply:
hmm... the only way I was able to get it working was by doing this:
```
import {Entity, model, property} from '@loopback/repository';
import {addCategoryPropertyMixin} from '../mixins/categoryPropertyMixin';

class BaseClass extends Entity {};

@model()
export class Note extends addCategoryPropertyMixin(BaseClass) {
  constructor(data?: Partial<Note>) {
    super(data);
  }

  @property({
    type: 'number',
    id: true,
    generated: true,
  })
  id?: number;

  @property({
    type: 'string',
    required: true,
  })
  title: string;

  @property({
    type: 'string',
  })
  content?: string;

}


export interface NoteRelations {
  // describe navigational properties here
}

export type NoteWithRelations = Note & NoteRelations;

```

Reply:
@raymondfeng suggested to me that we can make Entity a normal class (non-abstract)

Reply:
> @raymondfeng suggested to me that we can make Entity a normal class (non-abstract). Then I don't need to create BaseClass at all. Just pass in Entity.

+1, but I guess this will introduce some side effects.

I just remembered that i stumbled on this a year ago when using model mixins in my applications. I also introduced a `BaseModel` but with common attributes for all my models.

There is a similar issue when applying mixins on `ValueObject`.

Reply:
@derdeka , so you are ok with 

```
class BaseModel extends Entity {};

...

export class Note extends addCategoryPropertyMixin(BaseModel)

```

approach

and no longer ok with 

```
export class Note extends addCategoryPropertyMixin(Entity)
```
if it means making `Entity` non-abstract?

What kind of side-effects on Entity did this cause? I will run regression tests surrounding it to make sure there are no regressions (or side effects), but if you could elaborate, that would be helpful. thx.

Reply:
Both approaches are fine to me.